File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
//     vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        
//     }
    
vector<vector<int>> ans;
// set<vector<int>> s1;
void getsum(vector<int> &A, vector<int> arr, int i, long long s, int B){
    if(i>=A.size()||s>=B){
        if(s==B){
            // if(s1.find(arr)!=s1.end){
            ans.push_back(arr); 
            // s1.insert(arr);   
            // }
        }
        return;
    }
    getsum(A,arr,i+1,s,B);
    arr.push_back(A[i]);
    s+=A[i];
    getsum(A,arr,i,s,B);
    getsum(A,arr,i+1,s,B);
}

vector<vector<int> > combinationSum(vector<int> &A, int B) {
    
    sort(A.begin(),A.end());
    vector<vector<int>> tmp(0);
    ans = tmp;
    vector<int> arr(0);
    getsum(A,arr,0,0,B);
    set<vector<int>> s{ans.begin(),ans.end()};
ans.assign(s.begin(),s.end());
    sort(ans.begin(),ans.end(),[](const vector<int> &a1, const vector<int> &a2)->bool{
        int i=0,j=0;
        
        while(i<a1.size() && j<a2.size()){
            if(a1[i]<a2[j])return true;
            else if(a1[i]>a2[j])return false;
            i++;j++;
        }
        if(i>=a1.size())return true;
        return false;
    });
    return ans;
    
}
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    Set<List<Integer>> ans = new HashSet<>();
    
    void checkCombination(int[] candidates, int index, List<Integer> arr, int currSum, int target){
        if(currSum>target){
            return;
        }
        if(index>=candidates.length && currSum!=target){
            return;
        }
        if(currSum==target){
            ans.add(new ArrayList<>(arr));
            return;
        }
        // List<Integer> arr1 = new ArrayList<>(arr);
        checkCombination(candidates, index+1, arr,currSum,target);
        currSum+=candidates[index];
        // List<Integer> arr2 = new ArrayList<>(arr);
        arr.add(candidates[index]);
        checkCombination(candidates, index, arr,currSum,target);
        checkCombination(candidates, index+1, arr,currSum,target);
        arr.remove(arr.size()-1);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> arr = new ArrayList<>();
        checkCombination(candidates,0, arr, 0, target);
        List<List<Integer>> anslist = new ArrayList<>(ans);
        return anslist;
    }
}